corehq/apps/app_manager/static/app_manager/js/modules/report_module.js

```diff
@@ -318,6 +323,13 @@ hqDefine('app_manager/js/modules/report_module', function () {
                         break;
                     }
                 }
+
+                var duplicatedSlugs = getDuplicatedSlugs();
+                if (duplicatedSlugs.length !== 0) {
+                    alert(gettext("Report codes must be unique.  The following codes are duplicates: ")
+                          + duplicatedSlugs.join(", "));
+                }
```

Comment:
I used `alert` here because that's what the above validation check does, but that doesn't actually prevent saving - if the user dismisses the alert, the save goes through.  Is there another pattern for module validation in app builder?

Reply:
I think that generally we allow users to save their app no matter what, even if a build will fail (at least in form builder)

This looks like it only checks instances in the module. Is that a sufficient check? or should it be over app or domain? (We may have talked about this and I forgot the answer)

Reply:
You can use the `validate_for_build` methods on app manager models to do server-side validation and then add a message to build_errors.html. Grep for "no source module id" to see a simple example of such a validation.

Reply:
Also, for client-side validation, browser alerts aren't a good way to display errors (since the dialog takes control of the whole page, so you can't necessarily see where the error is), but I understand working with the pattern that's already on the page. I'll do a PR to clean it up.